File: getCueSubjects.py
#!/usr/bin/python

# script to return subject ids for subjects in cue fmri experiment. To use in a python script, do: 

		# from getCueSubjects import getsubs
		# subjects,gi = getsubs()

# returns subjects as a list of subject id strings and gi is a group index of group membership 
# (0 for controls, 1 for patients). Alternatively, do subjsA,_ = getsubs(0)	to return just 
# control subjects, or: subjsB,_ = getsubs(1) for just patients.   

import os,sys,socket,csv
from itertools import compress

#########  get main data directory and subjects to process	
def getMainDir():

	hostname=socket.gethostname()

	if hostname=='vta': 					# linux 
		main_dir='/home/span/lvta/cueexp'
	elif hostname=='sr15-9bb16ea2f9':		# Kelly's laptop
		main_dir='/Users/kelly/cueexp'
	else: 
		main_dir='/Users/kelly/cueexp'

	return main_dir


def getsubs(task='',group='all'):

	main_dir=getMainDir()

	# data directory
	data_dir=main_dir+'/data'
	#print 'data_dir:'+data_dir

	# define path to subject file
	subjFile = os.path.join(data_dir,'subjects_list','subjects_list.csv')
	

	## define variables to pull out of csv file
	subjects = [] # list of subject ids
	gi = []		  # list of corresponding group indices
		
	# these will be indices of which subjects to keep for each task	
	cue_idx = []
	mid_idx = []
	midi_idx = []
	dti_idx = []
	cue_sample1_idx = []
	cue_sample2_idx = []
	dti_mfb_idx = []
	relapse6mo_idx = []

	with open(subjFile) as csv_file:
		csv_reader = csv.reader(csv_file, delimiter=',')
		line_count = 0
		for line in csv_reader:
			if line_count == 0:
				colnames=line
				#print(colnames)
				line_count += 1
			else:
				subjects.append(line[0])
				gi.append(int(line[1]))
				cue_idx.append(bool(int(line[2])))
				mid_idx.append(bool(int(line[3])))
				midi_idx.append(bool(int(line[4])))
				dti_idx.append(bool(int(line[5])))
				cue_sample1_idx.append(bool(int(line[6])))
				cue_sample2_idx.append(bool(int(line[7])))
				dti_mfb_idx.append(bool(int(line[8])))

				# # deal with relapse NaN values
				# if (line[10][-3:]=='NaN')
				# relapse6mo_idx.append(bool(int(line[10])))

	# if a task string is given, return subset of subjects for that task
	if task=='cue': 
		subjects=list(compress(subjects,cue_idx))
		gi=list(compress(gi,cue_idx))

	elif task=='mid':
	
		subjects=list(compress(subjects,mid_idx))
		gi=list(compress(gi,mid_idx))

	elif task=='midi':
	
		subjects=list(compress(subjects,midi_idx))
		gi=list(compress(gi,midi_idx))

	elif task=='dti':
	
		subjects=list(compress(subjects,dti_idx))
		gi=list(compress(gi,dti_idx))

	elif task=='cue_sample1':
	
		subjects=list(compress(subjects,cue_sample1_idx))
		gi=list(compress(gi,cue_sample1_idx))

	elif task=='cue_sample2':
	
		subjects=list(compress(subjects,cue_sample2_idx))
		gi=list(compress(gi,cue_sample2_idx))

	elif task=='dti_mfb':
	
		subjects=list(compress(subjects,dti_mfb_idx))
		gi=list(compress(gi,dti_mfb_idx))

	# elif task=='relapse_6months':
	
	# 	subjects=list(compress(subjects,relapse6mo_idx))
	# 	gi=list(compress(gi,relapse6mo_idx))

	# elif task=='nonrelapse_6months':
	
	# 	subjects=list(compress(subjects,relapse6mo_idx))
	# 	gi=list(compress(gi,relapse6mo_idx))

	# if desired, return only controls or patients ids
	if str(group)=='controls' or str(group)=='0':
		idx = [i for i, x in enumerate(gi) if x == 0]
		gi = [gi[i] for i in idx]
		subjects = [subjects[i] for i in idx]
	
	if str(group)=='patients' or str(group)=='1':
		idx = [i for i, x in enumerate(gi) if x > 0]
		gi = [gi[i] for i in idx]
		subjects = [subjects[i] for i in idx]


	# return subjects and gi
	return subjects, gi


# def getsubs_claudia(task=''):

# 	# define subjects and gi lists
# 	subjects = [] # list of subject ids
# 	gi = []			# list of corresponding group indices

# 	# define path to subject file
# 	subjFile = os.path.join(os.path.expanduser('~'),'cueexp_claudia','data','subjects_list','subjects')
	
# 	with open(subjFile, 'r') as f:
# 		next(f) # omit header line
# 		for line in f:	
# 			subjects.append(line[0:line.find(',')])
# 			gi.append(line[line.find(',')+1:line.find(',')+3])
				
	
# 	# if a task string is given, return subset of subjects for that task
# 	if task: 

# 		omit_subs = [] # list of subjects to omit specific to this task
		
# 		omitSubjsFile = os.path.join(os.path.expanduser('~'),'cueexp_claudia','data','subjects_list','omit_subs_'+task)
	
# 		with open(omitSubjsFile, 'r') as f:
# 			next(f) # omit header line
# 			for line in f:	
# 				omit_subs.append(line[0:line.find(',')])
			
# 		for omit_sub in omit_subs:
# 			if omit_sub in subjects:
# 				gi.pop(subjects.index(omit_sub))
# 				subjects.remove(omit_sub)
	

# 	# make group indices integers
# 	gi = map(int,gi)		


# 	# return subjects and gi
# 	return subjects, gi



if __name__ == "__main__":
	subjects,gi = getsubs(sys.argv[1])
	print ' '.join(subjects)

#if __name__ == "__main__":
#	subjects,gi = getsubs(sys.argv[1])
#	print subjects
#	print gi


#	if __name__ == "__main__":
 #   	getsubs()
 # getsubs(int(sys.argv[1]))

File: getCueSubjects.m
function [subjects,gi] = getCueSubjects2(task,group)
% -------------------------------------------------------------------------
% [subjects,gi,notes,exc_subj_notes] = getCueSubjects(task,group)
% usage: returns cell array with subject id strings for this experiment.
% NOTE: this assumes that there is a file named 'subjects' within the exp
% data folder that has a list of the subject ids and a group index (0 for
% controls, 1 for patients).

% INPUT: 2 optional inputs:
%   task - string that must be either 'cue','mid', 'midi', or 'dti' or ''
%         (Default is '').
%   group - number or string specifying to return only subjects from a single group:
%         0 or 'controls' for control subs
%         1 or 'patients' for stimulant-dependent patients
%         'relapsers' for patient relapsers
%         'nonrelapsers' for patient nonrelapsers
%
% OUTPUT:
%   subjects - cell array of subject id strings for this experiment
%   gi(optional) - if desired, this returns a vector of 0s and 1s
%   indicating the group of the corresponding subject

% notes:
%

% author: Kelly, kelhennigan@gmail.com, 09-Nov-2014


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% return all subjects if task isn't given as input
if notDefined('task')
    task = '';
end

% make group '' by default, which means return all
if notDefined('group')
    group = '';
end

% get subjects_list directory
subjListFileName = fullfile(getCueBaseDir,'data','subjects_list','subjects_list.csv');

T=readtable(subjListFileName);


% define subject id cell array & vector of corresponding group indices
subjects = table2array(T(:,1));
gi=table2array(T(:,2));


% if task is defined, return only good subjects for that task
if ~isempty(task)
    
    taskindex = find(strcmpi(task,T.Properties.VariableNames));
    
    if isempty(taskindex)
        
        fprintf(['\ntask name ' task ' not recognized;\n returning all subjects...\n'])
        
    else
        
        subjects(table2array(T(:,taskindex))==0)=[];
        gi(table2array(T(:,taskindex))==0)=[];
    end
    
end


% now get only subjects from one specific group, if desired
if ~isempty(group)
    
    % return controls
    if strcmpi(group,'controls') || isequal(group,0)
        subjects = subjects(gi==0);
        gi = gi(gi==0);
        
        % return patients (return both VA and epiphany patients)
    elseif strcmpi(group,'patients') || isequal(group,1)
        subjects = subjects(gi>0);
        gi = gi(gi>0);
        
        % return patients with complete followup data (or confirmed relapse before then)
    elseif strcmpi(group,'patients_complete')
        ri=getCueData(subjects,'relapse');
        obs=getCueData(subjects,'observedtime');
        idx=find(ri==1 | obs>150); % either relapsed or followed up for >5 months
        subjects = subjects(idx);
        gi = gi(idx);
        
        % return patients with at least 3 months followup data (or confirmed relapse before then)
    elseif strcmpi(group,'patients_3months')
        ri=getCueData(subjects,'relapse');
        obs=getCueData(subjects,'observedtime');
        idx=find(ri==1 | obs>=90); % either relapsed or followed up for >=3 months
        subjects = subjects(idx);
        gi = gi(idx);
        
        % return relapsers
    elseif strcmpi(group,'relapsers')
        ri=getCueData(subjects,'relapse');
        subjects = subjects(ri==1);
        gi = gi(ri==1);
        
        
        % return nonrelapsers
    elseif strcmpi(group,'nonrelapsers')
        ri=getCueData(subjects,'relapse');
        subjects = subjects(ri==0);
        gi = gi(ri==0);
        
        % return those who relapsed within 3 mos
    elseif any(strcmpi(group,{'relapsers_3months','relapse_3months'}))
        ri=getCueData(subjects,'relapse_3months');
        subjects = subjects(ri==1);
        gi = gi(ri==1);
        
        
        % return those who did not relapse within 3 mos
    elseif any(strcmpi(group,{'nonrelapsers_3months','nonrelapse_3months'}))
        ri=getCueData(subjects,'relapse_3months');
        subjects = subjects(ri==0);
        gi = gi(ri==0);
        
        % return those who relapsed within 4 mos
    elseif any(strcmpi(group,{'relapsers_4months','relapse_4months'}))
        ri=getCueData(subjects,'relapse_4months');
        subjects = subjects(ri==1);
        gi = gi(ri==1);
        
        
        % return those who did not relapse within 4 mos
    elseif any(strcmpi(group,{'nonrelapsers_4months','nonrelapse_4months'}))
        ri=getCueData(subjects,'relapse_4months');
        subjects = subjects(ri==0);
        gi = gi(ri==0);
        
        % return those who relapsed within 6 mos
    elseif any(strcmpi(group,{'relapsers_6months','relapse_6months'}))
        ri=getCueData(subjects,'relapse_6months');
        subjects = subjects(ri==1);
        gi = gi(ri==1);
        
        
        % return those who did not relapse within 6 mos
    elseif any(strcmpi(group,{'nonrelapsers_6months','nonrelapse_6months'}))
        ri=getCueData(subjects,'relapse_6months');
        subjects = subjects(ri==0);
        gi = gi(ri==0);
        
        % return those who relapsed within 8 mos
    elseif any(strcmpi(group,{'relapsers_8months','relapse_8months'}))
        ri=getCueData(subjects,'relapse_8months');
        subjects = subjects(ri==1);
        gi = gi(ri==1);
        
        
        % return those who did not relapse within 8 mos
    elseif any(strcmpi(group,{'nonrelapsers_8months','nonrelapse_8months'}))
        ri=getCueData(subjects,'relapse_8months');
        subjects = subjects(ri==0);
        gi = gi(ri==0);
        
        
        % return only VA patients
    elseif strcmpi(group,'patients_for')
        subjects = subjects(gi==1);
        gi = gi(gi==1);
        
        
        % return only epiphany patients
    elseif strcmpi(group,'patients_epiphany')
        subjects = subjects(gi==2);
        gi = gi(gi==2);
        
        
    end  % group
    
end % if ~isempty(group)

end % function





